src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.Reflection.cs

```diff
@@ -0,0 +1,219 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System;
+using System.Reflection;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Microsoft.Extensions.DependencyInjection
+{
+    public static partial class ActivatorUtilities
+    {
+#if NET8_0_OR_GREATER // Use the faster ConstructorInvoker which also has alloc-free APIs when <= 4 parameters.
+        private static object ReflectionFactory_ServiceOnly_Fixed(
+            ConstructorInvoker invoker,
+            Type[] parameterTypes,
+            Type declaringType,
+            IServiceProvider serviceProvider)
+        {
+            Debug.Assert(parameterTypes.Length >= 1 && parameterTypes.Length <= FixedArgumentThreshold);
+            Debug.Assert(FixedArgumentThreshold == 4);
+
+            if (serviceProvider is null)
+                ThrowHelper_ArgumentNullException_serviceProvider();
+
+            object? arg1 = null;
+            object? arg2 = null;
+            object? arg3 = null;
+            object? arg4 = null;
+
+            switch (parameterTypes.Length)
+            {
+                case 4:
+                    arg4 = GetService(serviceProvider, parameterTypes[3], declaringType, false);
+                    goto case 3;
+                case 3:
+                    arg3 = GetService(serviceProvider, parameterTypes[2], declaringType, false);
+                    goto case 2;
+                case 2:
+                    arg2 = GetService(serviceProvider, parameterTypes[1], declaringType, false);
+                    goto case 1;
+                case 1:
+                    arg1 = GetService(serviceProvider, parameterTypes[0], declaringType, false);
+                    break;
+            }
+
+            return invoker.Invoke(arg1, arg2, arg3, arg4)!;
+        }
+
+        private static object ReflectionFactory_ServiceOnly_Span(
+            ConstructorInvoker invoker,
+            Type[] parameterTypes,
+            Type declaringType,
+            IServiceProvider serviceProvider)
+        {
+            ArgumentNullException.ThrowIfNull(nameof(serviceProvider));
+
+            object?[] arguments = new object?[parameterTypes.Length];
+            for (int i = 0; i < parameterTypes.Length; i++)
+            {
+                arguments[i] = GetService(serviceProvider, parameterTypes[i], declaringType, false);
+            }
+
+            return invoker.Invoke(arguments.AsSpan())!;
+        }
+
+        private static object ReflectionFactory_Canonical_Fixed(
+            ConstructorInvoker invoker,
+            FactoryParameterContext[] parameters,
+            Type declaringType,
+            IServiceProvider serviceProvider,
+            object?[]? arguments)
+        {
+            Debug.Assert(parameters.Length >= 1 && parameters.Length <= FixedArgumentThreshold);
+            Debug.Assert(FixedArgumentThreshold == 4);
+
+            if (serviceProvider is null)
```

Comment:
Similar to https://github.com/dotnet/runtime/pull/89573/files#r1278220352 , why are there so many ways of saying this? :)